File: Backend/Entities/Product.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace Virta.Entities
{
    public class Product
    {
        [Key]
        public Guid Id { get; set; }
        [Required]
        public string Title { get; set; }
        [Required]
        public decimal Price { get; set; }
        public string Description { get; set; }
        public string SKU { get; set; }
        [Required]
        public ProductTypes Type { get; set; }
        [Required]
        public ProductVisibilities Visibility { get; set; }
        [Required]
        public bool Active { get; set; }
        public virtual ICollection<Product> AssociatedProducts { get; set; }
        public virtual ICollection<ProductAttribute> ProductAttributes { get; set; }
        public virtual ICollection<Category> Categories { get; set; }
        public virtual ICollection<ProductImage> Images { get; set; }
        public virtual ICollection<Review> Reviews { get; set; }
        public ProductUnits Unit { get; set; }
        public decimal? OldPrice { get; set; }
        public ProductLabels Label { get; set; }
        public string Video { get; set; }

        public DateTime? CreatedAt { get; set; }
        public DateTime? UpdatedAt { get; set; }

        public enum ProductTypes
        {
            Simple, // AssociatedProducts => NULL
            Bundle, // AssociatedProducts => Bundled products
            Configurable  // AssociatedProducts => Product variations
        }

        public enum ProductVisibilities
        {
            Invisible, // Invisible
            PDP, // Only PDP
            PLP // PDP & PLP
        }

        public enum ProductLabels
        {
            None,
            New,
            Sale,
            Off,
            Featured,
            Trending
        }

        public enum ProductUnits
        {
            Piece,
            Gram,
            Liter,
            Meter,
            Kilogram,
            SquareMeter,
        }
    }
}

File: Frontend/src/app/models/Product.ts
import { CategoryDTO } from './Category';

export interface ProductDTO {
    id: string;
    description: string;
    sku: string;
    attributes: ProductAttributeDTO[];
    images: ProductImageDTO[];
    associatedProducts: ProductDTO[];
    title: string;
    price: number;
    type: ProductTypes;
    visibility: ProductVisibilities;
    active: boolean;
    categories: CategoryDTO[];
    stars: number;
    reviews: number;
    unit: ProductUnits;
    oldPrice: number;
    label: ProductLabels;
    video: string;
}

export interface ProductImageDTO {
    id: number;
    primary: boolean;
    url: string;
}

export interface ProductAttributeDTO {
    id: number;
    title: string;
    value: string;
    priority: number;
}

export enum ProductTypes
{
    Simple, // AssociatedProducts => NULL
    Bundle, // AssociatedProducts => Bundled products
    Configurable  // AssociatedProducts => Product variations
}

export enum ProductVisibilities
{
    Invisible, // Invisible
    PDP, // Only PDP
    PLP // PDP & PLP
}

export enum ProductLabels
{
    None, // No label
    New,
    Sale,
    Off,
    Featured,
    Trending
}

export enum ProductUnits
{
    Piece,
    Gram,
    Liter,
    Meter,
    Kilogram,
    SquareMeter
}

export interface PLP {
    products: ProductDTO[];
    totalCount: number;
}

export interface ProductInCart {
    productId: string;
    title: string;
    price: number;
    unit: ProductUnits;
    imageUrl: string;
    quantity: number;
}

export interface ProductInWishlist {
    productId: string;
    title: string;
    price: number;
    unit: ProductUnits;
    imageUrl: string;
}


// TODO: DELETE THIS
export interface Product {
    id: string;
    sku: string;
    title: string;
    price: number;
    images: ProductImage[];
    description?: string;
    attributes?: ProductAttribute[];
}

export interface ProductImage {
    id: number;
    primary: boolean;
    url: string;
}

export interface ProductAttribute {
    id: number;
    title: string;
    value: string;
    priority: number;
}

